File: Scripts/fuelsmart_data/units.py
"""Unit conversions used by the ingestion pipeline.

The bundled dataset stores one canonical set of units so the Swift calculation
engine never needs a per-country code path:

    liquid fuel consumption  ->  L/100 km
    electrical consumption   ->  kWh/100 km
    distance                 ->  km

U.S. source figures (MPG, kWh/100 mi) are converted here, and the original
official numbers are carried alongside for display so that a U.S. user still
sees the exact figure the EPA published rather than a round-tripped one.

These constants are exact definitions, not approximations.
"""

from __future__ import annotations

KM_PER_MILE = 1.609344
LITRES_PER_US_GALLON = 3.785411784
LITRES_PER_IMPERIAL_GALLON = 4.54609

# 100 km expressed in US gallons per mile terms: 100 * L/gal / km/mi
_US_MPG_TO_L100KM = 100.0 * LITRES_PER_US_GALLON / KM_PER_MILE  # ~235.2145
_IMP_MPG_TO_L100KM = 100.0 * LITRES_PER_IMPERIAL_GALLON / KM_PER_MILE  # ~282.4809


def us_mpg_to_l_per_100km(mpg: float | None) -> float | None:
    """Convert U.S. miles-per-gallon to litres per 100 km."""
    if mpg is None or mpg <= 0:
        return None
    return _US_MPG_TO_L100KM / mpg


def l_per_100km_to_us_mpg(l100: float | None) -> float | None:
    if l100 is None or l100 <= 0:
        return None
    return _US_MPG_TO_L100KM / l100


def imperial_mpg_to_l_per_100km(mpg: float | None) -> float | None:
    """Convert Imperial (Canadian) miles-per-gallon to litres per 100 km."""
    if mpg is None or mpg <= 0:
        return None
    return _IMP_MPG_TO_L100KM / mpg


def kwh_per_100mi_to_kwh_per_100km(kwh100mi: float | None) -> float | None:
    """EPA publishes electrical consumption per 100 miles."""
    if kwh100mi is None or kwh100mi <= 0:
        return None
    return kwh100mi / KM_PER_MILE


def miles_to_km(miles: float | None) -> float | None:
    if miles is None or miles <= 0:
        return None
    return miles * KM_PER_MILE


def grams_per_mile_to_grams_per_km(gpm: float | None) -> float | None:
    if gpm is None or gpm < 0:
        return None
    return gpm / KM_PER_MILE


def round_or_none(value: float | None, places: int) -> float | None:
    """Round for storage only. Intermediate calculation values are never rounded."""
    if value is None:
        return None
    return round(value, places)

File: Packages/FuelSmartCore/Sources/FuelSmartCore/Support/Units.swift
import Foundation

/// The region a comparison belongs to.
///
/// A comparison is always expressed wholly in one region's units. FuelSmart
/// never mixes units or shows a conversion in parentheses.
public enum Region: String, Codable, Sendable, CaseIterable, Hashable {
    case canada = "CA"
    case unitedStates = "US"

    public var currencyCode: String {
        switch self {
        case .canada: "CAD"
        case .unitedStates: "USD"
        }
    }

    /// Canada measures fuel use as litres per 100 km; the U.S. as miles per gallon.
    public var usesMetricDistance: Bool { self == .canada }
}

/// How the user prefers to type a distance. Everything is normalized to a year.
public enum DistancePeriod: String, Codable, Sendable, CaseIterable, Hashable {
    case day, week, month, year

    /// Periods per year. 365.25 absorbs leap years so a "per day" entry does not
    /// quietly differ from the same figure entered per year.
    public var periodsPerYear: Double {
        switch self {
        case .day: 365.25
        case .week: 52.1775
        case .month: 12
        case .year: 1
        }
    }
}

/// Exact unit conversions.
///
/// Every constant here is a definition, not a measurement, so these conversions
/// are lossless in both directions. The calculation engine works only in the
/// canonical units; conversion happens at the edges, on input and on display.
///
/// Canonical units: distance in kilometres, liquid fuel in litres per 100 km,
/// electricity in kWh per 100 km, money in the region's own currency.
public enum UnitConversionService {

    // MARK: - Definitions

    public static let kilometresPerMile = 1.609344
    public static let litresPerUSGallon = 3.785411784
    public static let litresPerImperialGallon = 4.54609

    /// 100 × litres-per-gallon ÷ km-per-mile. Multiply-or-divide by this to move
    /// between U.S. MPG and L/100 km — the relationship is reciprocal, not linear,
    /// which is precisely why the engine never stores MPG.
    public static let usMPGToLitresPer100km = 100.0 * litresPerUSGallon / kilometresPerMile

    public static let imperialMPGToLitresPer100km = 100.0 * litresPerImperialGallon / kilometresPerMile

    // MARK: - Distance

    public static func milesToKilometres(_ miles: Double) -> Double { miles * kilometresPerMile }
    public static func kilometresToMiles(_ km: Double) -> Double { km / kilometresPerMile }

    /// Convert a distance typed in the region's own unit into canonical kilometres.
    public static func distanceToKilometres(_ value: Double, region: Region) -> Double {
        region.usesMetricDistance ? value : milesToKilometres(value)
    }

    /// Convert canonical kilometres into the region's display unit.
    public static func kilometresToRegionDistance(_ km: Double, region: Region) -> Double {
        region.usesMetricDistance ? km : kilometresToMiles(km)
    }

    // MARK: - Fuel economy

    /// U.S. MPG → L/100 km. Returns nil for non-positive input, which is not a
    /// physically meaningful economy figure.
    public static func usMPGToLitresPer100km(_ mpg: Double) -> Double? {
        guard mpg > 0 else { return nil }
        return usMPGToLitresPer100km / mpg
    }

    public static func litresPer100kmToUSMPG(_ litresPer100km: Double) -> Double? {
        guard litresPer100km > 0 else { return nil }
        return usMPGToLitresPer100km / litresPer100km
    }

    public static func imperialMPGToLitresPer100km(_ mpg: Double) -> Double? {
        guard mpg > 0 else { return nil }
        return imperialMPGToLitresPer100km / mpg
    }

    // MARK: - Fuel price

    /// A price typed per U.S. gallon becomes a price per litre.
    public static func pricePerUSGallonToPerLitre(_ price: Double) -> Double {
        price / litresPerUSGallon
    }

    public static func pricePerLitreToPerUSGallon(_ price: Double) -> Double {
        price * litresPerUSGallon
    }

    /// Convert a fuel price typed in the region's own unit into canonical $/litre.
    public static func fuelPriceToPerLitre(_ value: Double, region: Region) -> Double {
        region.usesMetricDistance ? value : pricePerUSGallonToPerLitre(value)
    }

    public static func fuelPriceFromPerLitre(_ perLitre: Double, region: Region) -> Double {
        region.usesMetricDistance ? perLitre : pricePerLitreToPerUSGallon(perLitre)
    }

    // MARK: - Electricity

    /// The EPA publishes electrical consumption per 100 miles.
    public static func kWhPer100MilesToPer100km(_ value: Double) -> Double {
        value / kilometresPerMile
    }

    public static func kWhPer100kmToPer100Miles(_ value: Double) -> Double {
        value * kilometresPerMile
    }

    // MARK: - Emissions

    public static func gramsPerMileToGramsPerKilometre(_ value: Double) -> Double {
        value / kilometresPerMile
    }
}
